In [162]:
import numpy as np

# 3 punto

In [ ]:
def calculate_polynomial(matrix):
  '''
  Esta funcion realiza el calculo del polinomio caracteristico para una matriz cuadrada dada

  Parametros de entrada:
    matrix (numpy.ndarray): Descripción de matriz

  Retorna:
  numpy.ndarray: Descripción del valor retornado

  '''
  # Se calcula los eigenvalores de la matriz de entrada
  eigenvalores = np.linalg.eigvals(matrix)

  # Se retorna los eigenvalores de la matriz
  return eigenvalores

In [ ]:
matriz = np.array([[2, 1, 1],
                    [4, 3, 2],
                    [1, 1, 2]])

vals = calculate_polynomial(matriz)
print(vals)

[5.44948974 0.55051026 1.        ]


Luego de obtener los eigenvalores para lambda λ, se realiza el calculo:

(λ - 5.45)(λ - 0.550)(λ - 1) = 0


**Solución:**

λ³ - 7.000 λ² + 8.977 λ - 2.997 = 0

# 2 Punto

In [ ]:
from scipy.special import factorial

def euler(n):
  '''
  Esta funcion realiza el calculo del valor del número de Euler

  Parametros de entrada:
    n (int): Descripción de n

  Retorna:
  int: Descripción del valor retornado

  '''
  l = np.arange(n)
  factoriales = factorial(l)
  return np.sum(((1) / (factoriales) ) )

In [ ]:
euler(4)

2.6666666666666665

In [ ]:
euler(10)

2.718281525573192

In [ ]:
euler(10000)

2.718281828459045

In [ ]:
euler(100000000)

2.718281828459045


**Solución:**


El valor deseado se alcanza con la primera potencia de 10, es decir 10.

In [157]:
# Base de datos de productos y precios
base_de_datos = {
            "Pantalon": 100000,
            "Camiseta": 50000,
            "Gafas": 25000,
            "Gorra": 20000,
            "Zapatos": 250000
            # Agrega más productos y precios según sea necesario
        }


class Pedido:

    """
    Clase Pedido que ayuda a tomar los pedidos enviados por el usuario y calcula el total de la compra

    -----------

    Atributos
    ----------
    numero_pedido : int
        Identificador del pedido
    cliente : str
        Nombre del cliente quien realiza el pedido
    fecha : str
        Fecha de creación del pedido
    productos : dict
        Diccionario de productos seleccionados por el usuario


    Metodos
    -------
    mostrar_detalle()
        Muestra los productos y la cantidad de ellos

    agregar_producto(producto, cantidad)
        Agrega nuevos productos o actualiza los exitentes

    eliminar_producto(producto)
        Elimina el producto seleccionado

    calcular_total()
        Calcula el total de los productos realizando una busqueda de los precios en la base de datos
    """
    def __init__(self, numero_pedido: int, cliente: str, fecha: str, productos: dict):
      self.informacion_pedido = {
          'numero_pedido': numero_pedido,
          'cliente': cliente,
          'fecha': fecha,
          'productos': dict(productos),
          'total' : 0,
      }

      """
      Parametros
      ----------
      numero_pedido : int
          Identificador del pedido
      cliente : str
          Nombre del cliente quien realiza el pedido
      fecha : str
          Fecha de creación del pedido
      productos : dict
          Diccionario de productos seleccionados por el usuario
      """



    def mostrar_detalle(self):

      print(f"Número de Pedido: {self.informacion_pedido['numero_pedido']}")
      print(f"Cliente: {self.informacion_pedido['cliente']}")
      print(f"Fecha: {self.informacion_pedido['fecha']}")
      print("Productos:")
      for producto, cantidad in self.informacion_pedido['productos'].items():
          print(f" - {producto}: {cantidad}")
      self.calcular_total()
      print()
      print(f"Total: COP {self.informacion_pedido['total']}")

    def agregar_producto(self, producto, cantidad):
      #Se realiza la busqueda del producto en los anteriores productos, si lo encuentra suma la cantidad sino lo agrega
      if producto in self.informacion_pedido['productos']:
          self.informacion_pedido['productos'][producto] += cantidad
      else:
          self.informacion_pedido['productos'][producto] = cantidad

    def eliminar_producto(self, producto):
      #Valida si el producto se encuentra en los productos de ser asi lo elimina
      if producto in self.informacion_pedido['productos']:
          del self.informacion_pedido['productos'][producto]
      else:
          print(f"El producto '{producto}' no se encuentra en el pedido.")

    def calcular_total(self):
      total = 0
      for producto, cantidad in self.informacion_pedido['productos'].items():
          if producto in base_de_datos:
              precio = base_de_datos[producto]
              total += precio * cantidad
          else:
            print(f"*** El producto {producto} seleccionado por el usario no existe en la base de datos ***")
      self.informacion_pedido['total'] = total
      return total


In [158]:
# Crear un pedido
productos_iniciales = {"Pantalon": 1, "Camiseta": 1}
pedido1 = Pedido(1, "Laura", "2023-08-05", productos_iniciales)

In [159]:
# Mostrar detalles del pedido
pedido1.mostrar_detalle()

Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-05
Productos:
 - Pantalon: 1
 - Camiseta: 1

Total: COP 150000


In [160]:
pedido1.eliminar_producto("Pantalon")
pedido1.mostrar_detalle()


Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-05
Productos:
 - Camiseta: 1

Total: COP 50000


In [152]:
# Agregar productos
pedido1.agregar_producto("collar", 1)
# Mostrar productos nuevamente
pedido1.mostrar_detalle()

Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-05
Productos:
 - Pantalon: 1
 - Camiseta: 1
 - collar: 1
*** El producto collar seleccionado por el usario no existe en la base de datos ***

Total: COP 150000


In [155]:
class PedidoDescuento(Pedido):

      """
    Clase PedidoDescuento que ayuda a realizar un descuento de los productos seleccionados

    -----------

    Atributos
    ----------
    numero_pedido : int
        Identificador del pedido
    cliente : str
        Nombre del cliente quien realiza el pedido
    fecha : str
        Fecha de creación del pedido
    productos : dict
        Diccionario de productos seleccionados por el usuario
    descuento : int
        valor con el que se aplicará el descuento


    Metodos
    -------
    mostrar_detalle()
        Muestra los productos y la cantidad de ellos el total con y sin descuento
    """

    def __init__(self, numero_pedido: int, cliente: str, fecha: str, productos: dict, descuento: int):
        super().__init__(numero_pedido, cliente, fecha, productos)
        self.descuento = descuento

        """
        Atributos
        ----------
        numero_pedido : int
            Identificador del pedido
        cliente : str
            Nombre del cliente quien realiza el pedido
        fecha : str
            Fecha de creación del pedido
        productos : dict
            Diccionario de productos seleccionados por el usuario
        descuento : int
            valor con el que se aplicará el descuento
        """


    def mostrar_detalle(self):
        super().mostrar_detalle()
        super().calcular_total()
        num_aleatorio = random.random()
        #if num_aleatorio >= 0.5:
        descuento_total = self.informacion_pedido['total'] * (self.descuento / 100)
        self.informacion_pedido['total'] -= descuento_total
        #else:
          #print("Esta vez no ganaste el descuento")
        print(f"Descuento aplicado: {self.descuento}%")
        print(f"Total con descuento: COP {self.informacion_pedido['total']}")


In [156]:
pedido_descuento = PedidoDescuento(numero_pedido=1, cliente="Laura", fecha="2023-08-07", productos=productos_iniciales, descuento=50)

# Mostrar los detalles del pedido con descuento
pedido_descuento.mostrar_detalle()

Número de Pedido: 1
Cliente: Laura
Fecha: 2023-08-07
Productos:
 - Pantalon: 1
 - Camiseta: 1

Total: COP 150000
Descuento aplicado: 50%
Total con descuento: COP 75000.0


#Ejercicio 2

In [165]:
# Coeficientes de las ecuaciones
coefficients = np.array([[2, 3, -1, 4, 5],
                         [1, -2, 4, -3, 1],
                         [3, 2, -3, 5, -2],
                         [4, 1, -2, 3, 2],
                         [1, 1, 1, 1, 1]])

# Términos independientes
constants = np.array([15, 6, 11, 8, 7])

In [168]:
solution = np.linalg.solve(coefficients, constants)

print("Solución:\n")
print("x =", solution[0])
print("y =", solution[1])
print("z =", solution[2])
print("w =", solution[3])
print("v =", solution[4])



Solución:

x = -1.6774193548387097
y = -7.870967741935485
z = 5.129032258064516
w = 10.0
v = 1.4193548387096777
